### PR TITLE
Graph JSON framework

### PR DESCRIPTION
**Pull request by mbkumar**
This PR enables simsopt to properly emit and parse json documents for the simsopt objects which have graph dependencies. This PR incorporates changes in PR #260. Following are the changes implemented. 

1. Monty JSON parsers and emitters are moved into simsopt and modified for graph json framework.
2. The modified json related classes are renamed as GSONEncoder, GSONDecoder and GSONable.
3. Support for JAX is added in GSONEncoder. 
4. New wrapper class SIMSON, needed for graph json framework to work, is implemented.
5. All the classes that subclass from either Optimizable or GSONable are updated to work with the modified GSONEncoder and GSONDecoder classes.
6. JSON files in the examples folder are replaced with newer versions that support graph JSON.

### CODE REVIEW

**Review by landreman — COMMENTED** [review]
codecov is flagging many lines of `json.py` as non-tested. More test coverage of that file is needed.

There are many lines that are commented out, which would be better to remove.

Many of the routines for testing serialization and deserialization use `json.dumps` and `json.loads`. However users would not normally call those functions - users would instead use `.save()` and `.load()` functions, right? So does it make sense to change or add to the test routines to better cover the `.save()` and `.load()` functions that users would actually use?

**Review by landreman — COMMENTED** [review]
Looks almost ready.

Many of the routines for testing serialization and deserialization use `json.dumps` and `json.loads`. However users would not normally call those functions - users would instead use `.save()` and `.load()` functions, right? So how about changing some of the serialization/deserialization tests to use the `.save()` and `.load()` functions that users would actually use?

File: src/simsopt/geo/surfacerzfourier.py [hunk 1]
```diff
@@ -73,6 +73,17 @@ def __init__(self, nfp=1, stellsym=True, mpol=1, ntor=0,
...
+    def with_gridpoint_num(cls, nphi, ntheta, nfp=1, stellsym=True, mpol=1,
+                           ntor=0, range="full torus"):
+        quadpoints_phi, quadpoints_theta = Surface.get_quadpoints(nfp=nfp,
+                                                                  nphi=nphi,
+                                                                  ntheta=ntheta,
+                                                                  range=range)
+        return cls(nfp=nfp, stellsym=stellsym, mpol=mpol, ntor=ntor,
+                   quadpoints_phi=quadpoints_phi,
+                   quadpoints_theta=quadpoints_theta)
+
```
[the remark this answers is not in the document]
Comment (mbkumar) [reply src/simsopt/geo/surfacerzfourier.py:86]: Good catch. Deleted it.

**Comment by mbkumar** [comment]
@landreman 
Used save and load functions as well as methods in class to save and load in json format at couple of places.

**Review by landreman — APPROVED** [review]

**Pull request merged**